llvm/lib/Transforms/Scalar/LICM.cpp

```diff
@@ -2661,21 +2670,31 @@ static bool hoistAddSub(Instruction &I, Loop &L, ICFLoopSafetyInfo &SafetyInfo,
   return false;
 }
 
+static BinaryOperator *isReassociableOp(BinaryOperator *BO, unsigned Opcode1,
+                                        unsigned Opcode2) {
+  if (BO->getOpcode() == Opcode1 || BO->getOpcode() == Opcode2)
+    if (!isa<FPMathOperator>(BO) ||
+        (BO->hasAllowReassoc() && BO->hasNoSignedZeros()))
+      return BO;
+  return nullptr;
+}
+
```

Comment:
It is worth making the int and fp opcodes explicit to simplify the function like:
```
if (BO->getOpcode() == IntOpode)
  return BO;
if (BO->getOpcode() == FpOpcode && BO->hasAllowReassoc() && BO->hasNoSignedZeros())
  return B0;
return nullptr;
```